src/utils/mod.rs

```diff
@@ -673,3 +673,19 @@ pub fn camel_case_from(s: &str) -> usize {
     }
     last_i
 }
+
+/// Return whether a method returns `Self`.
+pub fn returns_self(cx: &LateContext, ret: &FunctionRetTy, ty: ty::Ty) -> bool {
+    if let FunctionRetTy::Return(ref ret_ty) = *ret {
+        let ast_ty_to_ty_cache = cx.tcx.ast_ty_to_ty_cache.borrow();
```

Comment:
Well, I just tried `ctxt::node_types` which returns `None` and `InferCtxt::node_ty{pe,}` which panic :sweat_smile:.